### ISSUE TITLE
@action decorator fail with typescript targeting ES3

### ISSUE THREAD

**Issue by testerez** [mobxjs/mobx#443]
The `@action` decorator seems to have no effect when compiled with typescript.
Here is a codepen: https://codepen.io/testerez/pen/qNKVvO

The generated JS shows that typescript at least tries to apply the decorator:
```
    __decorate([
        action
    ], Data.prototype, "increment");
```

Switching to babel fixes the issue. Is that a typescript issue?
I experience the same issue with typescript@beta (2.0)

**Comment by mweststrate** [comment]
This is really strange, because this is pretty extensively tested in the testsuite, but I can reproduce it in your codepen indeed. Does it only happen in codepen or also in your local setup if applicable?

**Comment by testerez** [comment]
Yes, it happened in my local setup at the first place (using ts@2)

**Comment by mweststrate** [comment]
Ok, this is really weird, if I compile the same code locally (using ts@2.0.0 as well), it compiles into:

```
 __decorate([
        action
    ], Data.prototype, "increment", null);
```

(note the `null` at the end, which forces another branch to be taken in the `__decorate` function). What does it compile to locally in your case? Did you enable the compiler flag `experimentalDecorators` ?

**Comment by andykog** [comment]
Seems the same problem as https://github.com/mobxjs/mobx/pull/343
Is your typescript target set to ES3 (default)?

**Comment by andykog** [comment]
Or are you using `reflect-metdata` package?
@testerez

**Comment by testerez** [comment]
In fact I was using the default ES3 target. Switching to ES5 solved the issue!

ES3 (ko)
```
	    __decorate([
	        mobx_1.action
	    ], Data.prototype, "increment");
```

ES5 (ok)
```
	    __decorate([
	        mobx_1.action
	    ], Data.prototype, "increment", null);
```

BTW, i'm discovering Mobx and enjoying it a lot! Thanks for this amazing work!

**Comment by andykog** [comment]
@testerez Great! I wouldn't close this issue yet, because ES3 target should work also.

**Comment by mweststrate** [comment]
es3 is not a supported target; it lacks object.defineProperty which is used extensively by MobX. (no ie8, can live with that). i am actually very suprised that TS doesn't refuse to compile with ES3 + decorators. I'll make sure to mention it explicitly in the docs

**Comment by andykog** [comment]
@mweststrate, but using `reflect-metdata` package with TypeScript targeting ES5 might lead to the same problem. Won't it be nicer to fix this instead of prohibiting ES3 target?

**Comment by mbriggs** [comment]
...and i think i just hit the issue @andykog was worried about (feel free to ask me to open a new ticket)

```
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var mobx_1 = require('mobx');
var ImageData = (function () {
    function ImageData() {
        this.initial = true;
        // @observable
        this.url = null;
        // @observable
        this.angle = 0;
        // @observable
        this.position = {
            x: null,
            y: null
        };
        // @observable
        this.scale = {
            x: 1,
            y: 1
        };
        // @observable
        this.dimensions = {
            width: -1,
            height: -1
        };
    }
    ImageData.prototype.reset = function () {
        this.url = null;
        this.angle = 0;
        this.position.x = null;
        this.position.y = null;
        this.scale.x = 1;
        this.scale.y = 1;
        this.dimensions.height = -1;
        this.dimensions.width = -1;
    };
    __decorate([
        mobx_1.observable, 
        __metadata('design:type', Boolean)
    ], ImageData.prototype, "initial", void 0);
    return ImageData;
}());
exports.ImageData = ImageData;
//# sourceMappingURL=image.js.map
```

and when I run tests, I get 

```

/Users/matt/src/layout-data/node_modules/mobx/lib/mobx.js:2209
                Object.defineProperty(target, key, descriptor);
                       ^
TypeError: Property description must be an object: undefined
    at Function.defineProperty (native)
    at classPropertyDecorator (/Users/matt/src/layout-data/node_modules/mobx/lib/mobx.js:2209:24)
    at observableDecorator (/Users/matt/src/layout-data/node_modules/mobx/lib/mobx.js:360:26)
    at observable (/Users/matt/src/layout-data/node_modules/mobx/lib/mobx.js:365:36)
    at DecoratePropertyWithoutDescriptor (/Users/matt/src/layout-data/node_modules/reflect-metadata/Reflect.ts:1084:13)
    at Object.decorate (/Users/matt/src/layout-data/node_modules/reflect-metadata/Reflect.ts:201:20)
    at __decorate (/Users/matt/src/layout-data/lib/image.ts:4:92)
    at ../../lib/image.ts:6:5
    at Object.<anonymous> (../../lib/image.ts:43:2)
    at Module._compile (module.js:413:34)
```

**Comment by mweststrate** [comment]
@mbriggs would you be able to provide a small repo or test snippet + the minimal amount of deps that causes this problem? Then I'll try to grok the root cause of this issue and add a regression test

**Comment by andykog** [comment]
@mweststrate, you can try `npm run full-test` in this branch https://github.com/andykog/mobx/tree/reflectDecorateProblem
That is a sad picture

**Comment by andykog** [comment]
@mbriggs, you are using `reflect-metadata`, aren't you?

**Comment by mbriggs** [comment]
@andykog yup, trying to mark something as angular2 `@Injectable`, which has `@observable` properties

**Comment by mbriggs** [comment]
@andykog @mweststrate so looks like this was fixed in mobx or reflect metadata. I couldnt repro in a standalone project, so tried copying versions used in package.json. still couldnt repo, so removed the ^ from versions and got it :) . Uploading the repo anyways, just in case it helps narrow other issues around this. https://github.com/mbriggs/mobx-issue

**Comment by andykog** [comment]
@mbriggs, you are right, that problem was fixed in #423

**Comment by stevefan1999** [comment]
I was running into the same problem, it is not fixed.
https://github.com/mobxjs/mobx-react/issues/512#issuecomment-400999084

But consider that ES3 is so ancient-technology now (you are looking for pre IE8 browsers), a fixing for this is maybe a futile effort.
